lib/ros_value.h

```diff
@@ -100,48 +104,139 @@ class RosValue {
   };
 
   Type getType() const {
-    return type;
+    return type_;
   }
 
-  // Constructors
-  RosValue(const Type type)
-    : type(type)
-    // , primitive_info({})
-  {
-    if (type == Type::object || type == Type::array) {
-      throw std::runtime_error("Cannot create an object or array with this constructor");
+ private:
+  template<class ReturnType, class IndexType, class ChildIteratorType>
+  class _const_iterator {
+   public:
+    bool operator==(const ChildIteratorType& other) const {
+      return index_ == other.index_;
+    }
+
+    bool operator!=(const ChildIteratorType& other) const {
+      return !(*this == other);
+    }
```

Comment:
> Overall, I really like this change. Don't have a ton of feedback, just some thoughts:
> 
> 1. Since we're storing the ROS messages in one large vector, is there an upper limit of the ROS messages that we can handle, since we're effectively storing all of them?
So we actually store an entire chunk of messages together, not just one individual message now. So if you want to just hold on to a single message, theres a bit of overhead there. In terms of the upper limit, it would just be the size of your memory.
> 2. How were you able to identify some of the slowdowns you document in your PR description?
It was mostly a process of going through a flamegraph of all the calls, and trying to remove the ones that took up the most of the programms runtime.
> 3. Do you think it's worth including a latency test in our CI pipeline?
It might be worth it, although it might be kind of weird setting some arbitrary performance limit (I.E. too high above the actual performance and you don't catch regressions, to low and a minor change could cause a failure.)

Reply:
Done